pystiche_papers/sanakoyeu_et_al_2018/_loss.py

```diff
@@ -1,18 +1,255 @@
+        return cast(torch.Tensor, self.predictor(enc))
+
+    def calculate_score(self, input_repr: torch.Tensor) -> torch.Tensor:
+        return torch.mean(
+            binary_cross_entropy_with_logits(
+                input_repr,
+                torch.ones_like(input_repr)
+                if self._target_distribution
+                else torch.zeros_like(input_repr),
+            )
+        )
+
+    def calculate_accuracy(self, input_repr: torch.Tensor) -> torch.Tensor:
+        comparator = torch.ge if self._target_distribution else torch.lt
+        return torch.mean(comparator(input_repr, 0.0).float())
+
+
+class MultiLayerPredictionOperator(ops.MultiLayerEncodingOperator):
+    r"""Convenience container for multiple :class:`PredictionOperator` s.
+
+    Args:
+        multi_layer_encoder: Multi-layer encoder.
+        layers: Layers of the ``multi_layer_encoder`` that the children operators
+            operate on.
+        get_encoding_op: Callable that returns a children operator given a
+            :class:`pystiche.enc.SingleLayerEncoder` extracted from the
+            ``multi_layer_encoder`` and its corresponding layer weight.
+        layer_weights: Weights of the children operators passed to ``get_encoding_op``.
+            If ``"sum"``, each layer weight is set to ``1.0``. If ``"mean"``, each
+            layer weight is set to ``1.0 / len(layers)``. If sequence of ``float``s its
+            length has to match ``layers``. Defaults to ``"mean"``.
+        score_weight: Score weight of the operator. Defaults to ``1.0``.
+    """
+
+    def discriminator_operators(self) -> Iterator["EncodingDiscriminatorOperator"]:
+        for op in self.operators():
+            if isinstance(op, EncodingDiscriminatorOperator):
+                yield op
+
+    def real(self, mode: bool = True) -> "MultiLayerPredictionOperator":
+        for op in self.discriminator_operators():
+            op.real(mode)
+        return self
+
+    def fake(self) -> "MultiLayerPredictionOperator":
+        return self.real(False)
+
+    def get_accuracy(self) -> torch.Tensor:
+        accuracies = torch.stack([op.accuracy for op in self.discriminator_operators()])
+        return torch.mean(accuracies)
+
+
+def prediction_loss(
+    impl_params: bool = True,
+    discriminator_multi_layer_encoder: Optional[MultiLayerEncoder] = None,
+    scale_weights: Union[str, Sequence[float]] = "sum",
+    score_weight: Optional[float] = None,
+) -> MultiLayerPredictionOperator:
+    r"""Discriminator loss from :cite:`SKL+2018`.
+
+    Capture image details at different scales with the ``predictors`` and sum up
+    all losses and accuracies on the different ``scales``.
+
+    Args:
+        impl_params: If ``True``, uses the parameters used in the reference
+            implementation of the original authors rather than what is described in
+            the paper.
+        discriminator_multi_layer_encoder: Trainable
```

Comment:
Every `MultiLayerEncoder` is trainable by default. The difference in this paper is that it doesn't have to be pre-trained, but will be trained. Still this is detail, that should not make it into this docstring.